go/pkg/cas/upload.go

```diff
@@ -589,6 +604,144 @@ func (u *uploader) uploadBatch(ctx context.Context, reqs []*repb.BatchUpdateBlob
 	})
 }
 
+// stream uploads the item using ByteStream service.
+//
+// If the blob is already uploaded, then the function returns quickly and
+// without an error.
+func (u *uploader) stream(ctx context.Context, item *uploadItem, updateCacheStats bool) error {
+	if err := u.semByteStreamWrite.Acquire(ctx, 1); err != nil {
+		return err
+	}
+	defer u.semByteStreamWrite.Release(1)
+
+	// Allocate resources.
+	enc := zstdEncoders.Get().(*zstd.Encoder)
+	fileBuf := u.fileIOBufs.Get().([]byte)
+	streamBuf := u.streamBufs.Get().([]byte)
+	defer func() {
+		enc.Close()
+		zstdEncoders.Put(enc)
+		u.fileIOBufs.Put(fileBuf)
+		u.streamBufs.Put(streamBuf)
+	}()
+
+	// Open the item.
+	r, err := item.Open()
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+
+	// TODO(nodir): implement per-RPC timeouts. No nice way to do it.
+	rewind := false
+	return u.withRetries(ctx, func(ctx context.Context) error {
+		// Do not rewind if this is the first attempt.
+		if rewind {
+			if _, err := r.Seek(0, io.SeekStart); err != nil {
+				return err
+			}
+		}
+		rewind = true
+
+		// Start the byte stream.
+		stream, err := u.byteStream.Write(ctx)
+		if err != nil {
+			return err
+		}
+
+		// Compress using an in-memory pipe. This is mostly to accomodate the fact
+		// that zstd package expects a writer.
+		// Note that using io.Pipe() means we buffer only bytes that were not uploaded yet.
+		pr, pw := io.Pipe()
+		enc.Reset(pw)
+
+		// Read from disk and make RPCs concurrently.
+		eg, ctx := errgroup.WithContext(ctx)
+		eg.Go(func() error {
```

Comment:
your suggestion becomes mandatory with conditional compression. Will do.

Reply:
I've extracted streaming part, which is the bulk of the code. `eg.Go` stayed somewhat the same, i.e. I didn't create sub-functions for them. partially because those functions aren't large, partially because moving them our would require carrying lots of state in params causing actually more code. Most of it is comments anyway.

BTW, I think GitHub is doing a poor job here. In my editor tab is configured to two spaces, while on GitHub one tab looks like 6 spaces? So indentation appears much larger than it actually is. In Gerrit it looks like 4 spaces. So if we were using Gerrit, I would look better :)

Reply:
Looks much better with ?ts=2

https://github.com/bazelbuild/remote-apis-sdks/pull/320/files/a5351ae12a5c3384b5d0296658af924210cb4811#diff-bb4a3d7f90b86fd97be5e865f3f0a58dfc1ed61e1f77ac269d3b36fad7bdd238

Reply:
FWIW I think extracting the streaming part is already good enough - and I wish Github would let me use 2/4 by default, but I couldn't figure a way that doesn't involve me tweaking the URL everything, *sigh*. (thanks for the tip!)